### PR TITLE
Make All OS tests run on GCP instances

### PR DESCRIPTION
**Pull request by alpar-t**
This PR makes the necesary adaptations to the tests  and adds a power shell script to
invoke the OS tests on GCP instances connected as CI workers.

Also noticed that logs were not being produced by the tests and that theses were not using log4j so fixed that too. 

One of the difficulties in working on theses tests was that the tests just stalled with no indication where the problem is. 
To ease with the debugging, after process explorer suggested that the tests are running some commands, we now have multiple timeouts: one for the tests ( which will generate a thread dump ) and one for individual commands ( that bails with the command being ran and output and error so far ) to make it easier to see what went wrong. 

The tests were blocking because apparently the pipes to the sub-process were not closing, thus the threads were blocking on them and we were blocking indefinitely on the join.  I'm not sure why this doesn't happen in vagrant, but we now properly deal with it.

### CODE REVIEW

**Comment by atorok** [comment]
Here's a build scan from my testing: https://scans.gradle.com/s/lpgxftanojdv2/tests 

@rjernst it seems we are running each test multiple times, but we can address that separately

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Archives.java [hunk 1]
```diff
@@ -275,26 +259,45 @@ public static void runElasticsearch(Installation installation, Shell sh) throws
...
-                sh.env.entrySet().stream()
-                    .map(entry -> "$processInfo.Environment.Add('" + entry.getKey() + "', '" + entry.getValue() + "'); ")
-                    .collect(joining()) +
-                "$processInfo.UseShellExecute = $false; " +
-                "$process = New-Object System.Diagnostics.Process; " +
-                "$process.StartInfo = $processInfo; " +
-                "$process.Start() | Out-Null; " +
-                "$process.Id;"
-            );
+            if (System.getenv("username").equals("vagrant")) {
```
Comment (pugnascotia) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/Archives.java]: Most of the PowerShell lines in the following `sh.run()` calls are the same. Is there any value in reducing the duplication here?

**Review by pugnascotia — APPROVED** [review]
I can't say anything about the PowerShell, but the rest looks good.

**Review by mark-vieira — COMMENTED** [review]
Going to defer to @rjernst for the most part here since he's most familiar with the packaging test infrastructure.

**Comment by mark-vieira** [comment]
> This PR makes the necesary adaptations to the tests and adds a power shell script to
invoke the OS tests on GCP instances connected as CI workers.

Just to confirm, the build isn't actually spinning up workers. What we intend to do is launch a Windows CI work, then directly invoke the `destructive` test task on that worker instead of inside a Vagrant VM. Yes?

File: .ci/os.ps1 [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+#
+# $env:ES_BUILD_JAVA="java12" 
+#$env:ES_RUNTIME_JAVA="java11" 
+
+$ErrorActionPreference="Stop"
+$gradleInit = "C:\Users\$env:username\.gradle\init.d\"
+Remove-Item -Recurse -Force $gradleInit -ErrorAction Ignore
+New-Item -ItemType directory -Path $gradleInit
+
+# Copy-Item .ci/init.gradle -Destination $gradleInit
```
[the remark this answers is not in the document]
Comment (atorok) [reply .ci/os.ps1]: Thanks, that's left over from testing, I will un-comment it.

**Comment by atorok** [comment]
> Just to confirm, the build isn't actually spinning up workers. What we intend to do is launch a Windows CI work, then directly invoke the `destructive` test task on that worker instead of inside a Vagrant VM. Yes?

That's right. We will have a matrix job so CI will spin up the workers and invoke the  `ps1` script from this PR on each.

**Comment by atorok** [comment]
run @elasticsearch-ci/packaging

**Comment by atorok** [comment]
Windows test on CI workers passed : https://gradle-enterprise.elastic.co/s/p4qyrkswxipm4

**Comment by atorok** [comment]
CentOS passed: https://gradle-enterprise.elastic.co/s/dai35e6bvhk2m

File: qa/os/src/test/java/org/elasticsearch/packaging/test/ArchiveTests.java [hunk 3]
```diff
@@ -413,4 +408,11 @@ public void test94ElasticsearchNodeExecuteCliNotEsHomeWorkDir() throws Exception
         }
     }
 
+    @After
+    public void tearDown() throws Exception {
+        // clean up logs between logs so we only show logs from the running tests if we dump them on failure
+        FileUtils.rm(installation.logs);
```
Comment (williamrandolph) [inline qa/os/src/test/java/org/elasticsearch/packaging/test/ArchiveTests.java]: Would it make sense to move this to the parent class (`PackagingTestCase`) so that avoid having our logs carry over for all test groupings?

**Comment by atorok** [comment]
Debian also passing with the last fix: https://gradle-enterprise.elastic.co/s/xrc6baytgnqqg

**Comment by atorok** [comment]
@elasticmachine run elasticsearch-ci/2

File: .ci/os.ps1 [hunk 4]
```diff
@@ -0,0 +1,36 @@
+If (-NOT ([Security.Principal.WindowsPrincipal] [Security.Principal.WindowsIdentity]::GetCurrent()).IsInRole([Security.Principal.WindowsBuiltInRole]::Administrator))
+{
+    # Relaunch as an elevated process:
+    Start-Process powershell.exe "-File",('"{0}"' -f $MyInvocation.MyCommand.Path) -Verb RunAs
+    exit
+}
+
+# CI configures these, uncoment if running manually 
+#
+# $env:ES_BUILD_JAVA="java12"
```
Comment (rjernst) [inline .ci/os.ps1]: Can we just set these if they are not set so manual editing is not necessary?

File: .ci/os.ps1 [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+$env:PATH="C:\Users\jenkins\.java\$env:ES_BUILD_JAVA\bin\;$env:PATH"
+$env:JAVA_HOME=$null
+$env:SYSTEM_JAVA_HOME="C:\Users\jenkins\.java\$env:ES_RUNTIME_JAVA"
+Remove-Item -Recurse -Force \tmp -ErrorAction Ignore
+New-Item -ItemType directory -Path \tmp
+
+$ErrorActionPreference="Continue"
+# TODO: remove the task exclusions once dependencies are set correctly and these don't run for Windows or buldiung the deb on windows is fixed
+& .\gradlew.bat -g "C:\Users\$env:username\.gradle" --parallel --scan --console=plain destructiveDistroTest `
+   -x :distribution:packages:buildOssDeb `
```
Comment (rjernst) [inline .ci/os.ps1]: What is the task dep causing these to be built? We should never even try building these if they won't be used by the test, which was a major goal of the refactoring work here.

File: qa/os/src/test/java/org/elasticsearch/packaging/test/PackageTests.java [hunk 6]
```diff
@@ -121,20 +121,15 @@ public void test32JavaHomeOverride() throws Exception {
...
-            startElasticsearch(sh);
-            runElasticsearchTests();
-            stopElasticsearch(sh);
-        } finally {
-            if (unlinked) {
-                sh.run("ln -sf " + readlink.stdout.trim() + " /usr/bin/java");
-            }
+        // we don't require java be installed for the tests, but remove it if it's there
+        // since we don't require it for the tests, don't bother restoring it
+        if (Files.exists(Paths.get("/usr/bin/java"))) {
```
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/test/PackageTests.java]: Why was the previous code causing problems? If we are going to ensure /usr/bin/java doesn't exist, we should do so outside of test execution. This seems like an environmental issue, not something a test (in the middle of other tests running) should be changing. Deleting and not restoring system files is a serious, unexpected side effect to a test running.

File: qa/os/src/test/java/org/elasticsearch/packaging/util/FileUtils.java [hunk 7]
```diff
@@ -179,6 +181,30 @@ public static String slurpAllLogs(Path logPath, String activeLogFile, String rot
         }
     }
 
+    public static void logAllLogs(Path logsDir, Logger logger) {
+        if (Files.exists(logsDir) == false) {
+            logger.warn("Can't show logs from directory {} as it doesn't exists", logsDir);
+            return;
+        }
+        logger.info("Showing contents of directory: {} ({})", logsDir, logsDir.toAbsolutePath());
+        try(Stream<Path> fileStream = Files.list(logsDir)) {
```
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/FileUtils.java]: nit: space after try

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java [hunk 8]
```diff
@@ -68,6 +76,20 @@ public Result runIgnoreExitCode(String script) {
         return runScriptIgnoreExitCode(getScriptCommand(script));
     }
 
+    public void chown(Path path) throws Exception {
+        Platforms.onLinux(() -> run("chown -R elasticsearch:elasticsearch " + path));
+        Platforms.onWindows(() -> run(
+            "$account = New-Object System.Security.Principal.NTAccount '" + System.getenv("username")  + "'; " +
```
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java]: Don't we have the constant ARCHIVE_USER for use here and above?

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java [hunk 9]
```diff
@@ -91,68 +113,114 @@ public Result run( String command, Object... args) {
...
+        for (Map.Entry<String, String> entry : env.entrySet()) {
+            builder.environment().put(entry.getKey(), entry.getValue());
         }
-
+        final Path stdOut;
+        final Path stdErr;
         try {
+            Path tmpDir = Paths.get(System.getProperty("java.io.tmpdir"));
+            Files.createDirectories(tmpDir);
+            stdOut = Files.createTempFile(tmpDir, getClass().getName(), ".out");
```
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java]: Where are these files cleaned up?

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java [hunk 10]
```diff
@@ -91,68 +113,114 @@ public Result run( String command, Object... args) {
...
 
+    private String readFileIfExists(Path path) throws IOException {
+        if (Files.exists(path)) {
+            long size = Files.size(path);
+            if (size > 100 * 1024) {
+                return "<<Too large to read: " + size  + " bytes>>";
+            }
+            LinkedList<String> result = new LinkedList<>();
+            AtomicBoolean linesDiscarded = new AtomicBoolean(false);
+            try(Stream<String> lines = Files.lines(path, StandardCharsets.UTF_8)) {
```
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java]: nit: space after try

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java [hunk 11]
```diff
@@ -91,68 +113,114 @@ public Result run( String command, Object... args) {
...
+                lines.forEach(line -> {
+                    result.add(line);
+                    if (result.size() >= TAIL_WHEN_TOO_MUCH_OUTPUT) {
```
Comment (rjernst) [inline qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java]: Why are we omitting lines at all? Once the VM is destroyed we can't get them out. If we want to debate about the merits of what is output fine, but I think it should be a separate PR from this already large PR to get the OS tests running in gcp?

**Review by rjernst — COMMENTED** [review]
I left some comments

File: .ci/os.ps1 (hunk 4 above)
Comment (atorok) [reply .ci/os.ps1]: I was planning to read  `.ci/java-versions.properties` like in Linux, but in a follow up.

File: .ci/os.ps1 (hunk 5 above)
Comment (atorok) [reply .ci/os.ps1]: I'm not sure, but has to be something relating to the OS tests. 
I assumed the refactoring is not yet  complete. 
We built these before because we check that these can be extracted but I already addressed that.

Either way I would prefer to look at this in a separate PR.

File: qa/os/src/test/java/org/elasticsearch/packaging/test/PackageTests.java (hunk 6 above)
Comment (atorok) [reply qa/os/src/test/java/org/elasticsearch/packaging/test/PackageTests.java]: Vagrant VMs have a java package installed whereas CI workers do not. 
Since we label these as destructive and always run them on ephemeral VMs,  the restoring of this made the test unnecessarily hard to read.  I think it's worth restoring when it makes it possible to re-use a VM e.x. for development, but in this case, the test works if there's no java installed as well. 

The right fix here is probably to make sure java is not installed on the vagrant images ?

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java (hunk 8 above)
Comment (atorok) [reply qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java]: There's `ARCHIVE_OWNER` but that's part of the `ArchiveTests` and has hard coded values.

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java (hunk 9 above)
Comment (atorok) [reply qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java]: Good catch, they were not. I added cleanup.

File: qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java (hunk 11 above)
Comment (atorok) [reply qa/os/src/test/java/org/elasticsearch/packaging/util/Shell.java]: I'm going to remove omitting the lines, but will keep the size check.
One of the windows installer tests is generating a 16GB stderr.
I don't know what's in it because I didn't download it and nothing on windows could open it...

File: qa/os/src/test/java/org/elasticsearch/packaging/test/PackageTests.java (hunk 6 above)
Comment (atorok) [reply qa/os/src/test/java/org/elasticsearch/packaging/test/PackageTests.java]: I added the restore back to make this less awkward.

**Review by rjernst — APPROVED** [review]
LGTM, given the followup work you noted in response to some previous comments.
